src/lib/io/sol-i2c-impl-contiki-qmsi.c

```diff
@@ -0,0 +1,476 @@
+
+    xfer.tx = (uint8_t *)tx;
+    xfer.tx_len = tx_len;
+    xfer.rx = (uint8_t *)rx;
+    xfer.rx_len = rx_len;
+    xfer.id = id;
+    xfer.stop = stop;
+    xfer.tx_callback = tx_callback;
+    xfer.rx_callback = rx_callback;
+    xfer.err_callback = err_callback;
+
+    return qm_i2c_master_irq_transfer(i2c, &xfer, slave);
+}
+
+SOL_API struct sol_i2c *
+sol_i2c_open_raw(uint8_t bus, enum sol_i2c_speed speed)
+{
+    struct sol_i2c *i2c;
+    qm_i2c_config_t cfg;
+    qm_i2c_speed_t bus_speed;
+    qm_rc_t ret;
+
+    if (bus >= QM_I2C_NUM) {
+        SOL_WRN("I2C bus #%" PRIu8 " doesn't exist.", bus);
+        return NULL;
+    }
+
+    if ((i2c = buses[bus]) != NULL)
+        return i2c;
+
+    switch (speed) {
+    case SOL_I2C_SPEED_10KBIT:
+    case SOL_I2C_SPEED_100KBIT:
+        bus_speed = QM_I2C_SPEED_STD;
+        break;
+    case SOL_I2C_SPEED_400KBIT:
+        bus_speed = QM_I2C_SPEED_FAST;
+        break;
+    case SOL_I2C_SPEED_1MBIT:
+    case SOL_I2C_SPEED_3MBIT_400KBIT:
+        bus_speed = QM_I2C_SPEED_FAST_PLUS;
+        break;
+    default:
+        SOL_WRN("Unsupported speed value: %d", speed);
+        return NULL;
+    };
+
+    switch (bus) {
+    case QM_I2C_0:
+        qm_irq_request(QM_IRQ_I2C_0, qm_i2c_0_isr);
+        clk_periph_enable(CLK_PERIPH_CLK | CLK_PERIPH_I2C_M0_REGISTER);
+        break;
+#if QUARK_SE
+    case QM_I2C_1:
+        qm_irq_request(QM_IRQ_I2C_1, qm_i2c_1_isr);
+        clk_periph_enable(CLK_PERIPH_CLK | CLK_PERIPH_I2C_M1_REGISTER);
+        break;
+#endif
+    case QM_I2C_NUM:
+        /* We checked if we were passed the limit before, so we should never
+         * hit this point. Using all the enum values and no default, however,
+         * allows us to rely on the compiler to know if there are values
+         * we are not considering (depending on warning levels) */
+        break;
+    }
+
+    i2c = calloc(1, sizeof(*i2c));
+    SOL_NULL_CHECK(i2c, NULL);
+
+    i2c->bus = bus;
+
+    ret = qm_i2c_get_config(i2c->bus, &cfg);
+    SOL_EXP_CHECK_GOTO(ret != QM_RC_OK, error);
+
+    cfg.speed = bus_speed;
+    cfg.address_mode = QM_I2C_7_BIT;
+    cfg.mode = QM_I2C_MASTER;
+    cfg.slave_addr = 0;
+
+    ret = qm_i2c_set_config(i2c->bus, &cfg);
+    SOL_EXP_CHECK_GOTO(ret != QM_RC_OK, error);
+
+    if (!i2c_irq_event) {
+        bool r;
+
+        i2c_irq_event = process_alloc_event();
+        r = sol_mainloop_contiki_event_handler_add(&i2c_irq_event, NULL,
+            i2c_cb_dispatch, NULL);
+        SOL_EXP_CHECK_GOTO(!r, error);
+    }
+
+    buses[i2c->bus] = i2c;
+
+    return i2c;
+
+error:
+    free(i2c);
+    return NULL;
+}
+
+SOL_API void
+sol_i2c_close_raw(struct sol_i2c *i2c)
+{
+    SOL_NULL_CHECK(i2c);
+
+    buses[i2c->bus] = NULL;
+    free(i2c);
+}
+
+SOL_API bool
+sol_i2c_set_slave_address(struct sol_i2c *i2c, uint8_t slave_address)
+{
+    SOL_NULL_CHECK(i2c, false);
+
+    if (qm_i2c_get_status(i2c->bus) != QM_I2C_IDLE)
+        return false;
```

Comment:
On Thu, 19 May 2016 07:28:20 -0700, Gustavo Lima Chaves wrote:
> > +SOL_API void
> > +sol_i2c_close_raw(struct sol_i2c *i2c)
> > +{
> > +    SOL_NULL_CHECK(i2c);
> > +
> > +    buses[i2c->bus] = NULL;
> > +    free(i2c);
> > +}
> > +
> > +SOL_API bool
> > +sol_i2c_set_slave_address(struct sol_i2c *i2c, uint8_t slave_address)
> > +{
> > +    SOL_NULL_CHECK(i2c, false);
> > +
> > +    if (qm_i2c_get_status(i2c->bus) != QM_I2C_IDLE)
> > +        return false;
> 
> What if the slave address was per transaction? This would leave this idle wait open.

What?